### ISSUE TITLE
More Raid levels

### ISSUE THREAD

**Issue by NicholasRush** [openzfs/zfs#14741]
Hello everyone,

i have seen that in the last year the expansion feature of raidz/2/3 is done. This is a great feature but it needs to rewrite all the allocation information on all disks when drives added to that vdev. If you implement an Raid4 Raid-Type maybe Raid4z-1/-2/-3 with in max 3 parity drives, Then there is no need for it, because if an disk is added only all new writes are go to all old and the new disk. All old data needs to be rewritten from time to time to use the new added disk. That means you can use much faster the new added capacity than in raidz like arrays where the parity blocks rotate between all disks.

If you use ZFS at home as dataserver this could be a good innovation. But like all Raid4 arrays in the past the parity drives would be the bottlenek of the array this could be accomplished if the use of a `log` device  would be forced to always write full stripes including the parity.

**Comment by amotin** [comment]
I have feeling that you lack some understanding of how ZFS and its RAIDZ work.

I don't see how fixed parity position would allow faster use of the added space. Actually in opposite, without existing data replacement there would be some holes in every strip, that would be impossible to use until the existing data are deleted/overwritten.

Also now in RAIDZ each block has its own parity, not depending on anything else.  Use of fixed parity would either create a write hole similar to RAID5, or limited minimal block size similar to DRAID.

**Comment by NicholasRush** [comment]
> I have feeling that you lack some understanding of how ZFS and its RAIDZ work.
> 
> I don't see how fixed parity position would allow faster use of the added space. Actually in opposite, without existing data replacement there would be some holes in every strip, that would be impossible to use until the existing data are deleted/overwritten.
> 
> Also now in RAIDZ each block has its own parity, not depending on anything else. Use of fixed parity would either create a write hole similar to RAID5, or limited minimal block size similar to DRAID.

Im not a native us-english speaker/writer and i don't often need to write english. That could be a reason why it's could be heavy to understand what i want with my request. :) First sorry for that.

I use ZFS at work and in my homeserver for more than 10 years now. In many configurations `mirror`, `raidz` and `raidz3` for drives with 12 disks that will be bigger than 6TB. But at work i have used Netapp Ontap storage at the same time. And there i have seen the advantages and disadvantages of both systems. ZFS and there WAFL. At the time where ZFS was released by SUN it was the outstanding  technology of all cow filesystems on the market at that time. But today where ZFS is opensource it is also a great filesystem but it lacks all features that other vendors have. First there were the Block Pointer Rewrite Feature, which is a working feature of other vendors and would be also for ZFS a key technology to archive a better storage efficency. 

To now come to my intention to intergrate other Raidlevels like Raid4 with dual and tripple parity. 
A integration of that type of Raidlevel will have no writehole at all, if you always write full stripes including the parity and journal that on an external `LOG` device. But the algortihm of the last parity disk should not be xor like the first two, it should be an error correction algorithm like  Erasure Coding. To get forward error correction at raidlevel here.

If you add then more disks to that raid implementation, there is no need to have of a block pointer traversal first, because of the fixed block-size. The block size is then the size that you specify by the creation of a pool. The disadvantage of that is the same that draid has, you first have to resilver and later you have to scrub. But you can easy expand the vdev which is with draid not possible and change the parity level inplace.
To accomplish the paddings in a written stripe you could add a technique like netapp does with data compacting at the raid-level. It compacts physical blocks that are not complete filled together in one physical block for a better space usage. Thats a technique that could be also done at any scrub of the pool.

I hope this describes my intention better as before. It would be an innovation for all users that expands their arrays from time to time without the need to have all drives at pool creation.

**Comment by amotin** [comment]
> A integration of that type of Raidlevel will have no writehole at all, if you always write full stripes including the parity and journal that on an external `LOG` device.

You still do not understand how RAIDZ/DRAID and ZIL work.  ZIL logs user data, it has nothing to do with RAIDZ or DRAID.  RAIDZ and DRAID by definition have no write hole, since they always write full blocks.  You are proposing to add another log device to duplicate ALL writes to the data disks.  Excuse me, but there is no so fast log devices.  Such technique is usually implemented by battery-backed cache of hardware RAID controller.  Attempt to do it in software makes whole idea pretty pointless IMO.

**Comment by NicholasRush** [comment]
>> A integration of that type of Raidlevel will have no writehole at all, if you always write full stripes including the parity and journal that on an external LOG device.

I'm sorry I expressed myself wrongly here. Of course you are right that the "LOG" device only "logs" the synchronous user data and file system transactions. And that it is a device that is only written to and only read after a power failure or crash.

> You still do not understand how RAIDZ/DRAID and ZIL work. ZIL logs user data, it has nothing to do with RAIDZ or DRAID. RAIDZ and DRAID by definition have no write hole, since they always write full blocks. You are proposing to add another log device to duplicate ALL writes to the data disks. Excuse me, but there is no so fast log devices. Such technique is usually implemented by battery-backed cache of hardware RAID controller. Attempt to do it in software makes whole idea pretty pointless IMO.

I don't want an additional log device here and no duplicate write processes on it. But it makes sense not to use the "Zil" area on the ZFS pool with a Raid 4-like vdev in order not to slow down the i/o processes. And the "LOG" device should only be used in this case to record the operations on the file system and the drives and not the result.  That was my only thought behind it.

What I would like to suggest is the integration of an additional raid level similar to Raid4. In this Raid-4 implementation, each 4KB block on the parity disk is composed of a 4KB block from each of the data disks. In fact the stripe size is always 4K here. The parity on the parity hard disk is calculated here with XOR. A second parity disk is also XORed over the first parity drive and all but one data drive. With the third parity hard disk, the parity is calculated anti-diagonally using the triple erasure coding method. All data hard drives and the first parity hard drive are used for the calculation. The second parity disk is not included in the calculation.

The diagonal parity calculation algorithm requires that the sum of the number of all data drives plus the first drive with parity data is prime p and greater than 2. The excess and non-existent drives are assumed to be drives containing only zeros, the XOR parity calculation is not affected by zeros. This means that such a storage array can be easily expanded by adding additional empty drives without recalculating parities.

**Comment by amotin** [comment]
We are still not getting in sync.  If proposed RAID4 stripe is not equal to ZFS data block, i.e. there are several ZFS data blocks fit into the same stripe, then to update parity you'd need to read either old data and parity or data on all other disks.  It creates a write hole, that requires a lot of logging traffic, plus plenty of read-modify-write's and additional caching layer completely different from ARC.  Instead of diving into parity calculation details, please focus on those issues, otherwise we have nothing to talk about.  If the RAID4 stripe is supposed to always store only single ZFS block, then we are getting another form of DRAID, unable to store small data blocks and so having very limited usage, and it should be dramatically better that one to even bother.

**Comment by NicholasRush** [comment]
> We are still not getting in sync. If proposed RAID4 stripe is not equal to ZFS data block, i.e. there are several ZFS data blocks fit into the same stripe, then to update parity you'd need to read either old data and parity or data on all other disks. It creates a write hole, that requires a lot of logging traffic, plus plenty of read-modify-write's and additional caching layer completely different from ARC. Instead of diving into parity calculation details, please focus on those issues, otherwise we have nothing to talk about. If the RAID4 stripe is supposed to always store only single ZFS block, then we are getting another form of DRAID, unable to store small data blocks and so having very limited usage, and it should be dramatically better that one to even bother.

However, what you are addressing here is the case that is always present when you have a logging transaction-based file system that never overwrites anything. If I have to write a 1KB file with metadata and then write the same file again later, as mail servers do all the time, then you're wasting physical storage space. Because the written strip then contains only a small amount of user data and a lot of filler data. But you could possibly solve this circumstance for draid and this concept together. Raidz is also affected by this problem when small data is written. It's true though that it's mostly a problem for raid types with fixed block sizes.

The integration with ZFS, in order to make this possible, would always be to buffer the user data and transactions in the RAM in connection with the ZLOG for such raid types, but for a maximum of 15 seconds or until the ZLOG is full, so that everything is always complete strips with as much user data as possible are written. How else could you merge multiple writes and connect them together. In this way, all random write processes will automatically become sequential write processes.

If the data is then longer on the raid and the change has led to the fragmentation increasing, the only possibility at present, as with Raidz, is to carry out a zfs send and zfs receive so that the data is rewritten cleanly.

In my opinion, the only way to solve this in the long term is to implement an abstraction layer that controls the physically available blocks from the raid layer, so that small, contiguous blocks can then be merged into individual sequential strips. And that with these raid types you should always try to write complete strips with parity. Complete here means filled with user data.

With an abstraction layer, it would also be possible to carry out a limited block pointer rewrite on the raid layer and subsequently merge stripes with filler data into a stripe without filler data during a scrub. But that would have to be integrated into ZFS in such a way that communication through the storage layers works.

It was often desired to introduce an additional checksum method, which would then also be possible here.

Without tight integration, it would be nothing more than ZFS on an MDRAID where no layer knows anything about each other.

**Comment by NicholasRush** [comment]
I have never used the separation of metadata via special devices because, as far as I know, there is no way to remove them from the pool and get the metadata back on the spindles.

**Comment by NicholasRush** [comment]
After much research in this area, I would like to return to this topic.

Especially because Raid-Z expansion has been released in the latest OpenZFS version and many people are asking about the parity extension Raid-Z1 to Raid-Z2 or Raid-Z3.

If the RAID-4 type is implemented in OpenZFS, which implements XOR on the parity disks in linear diagonal and anti-diagonal parity, the entire expansion problem is completely solved. RAID-4 functions like RAID-0, but with parity disks. The OpenZFS logical block is written to the data disks in one go with the checksum, and the corresponding checksums are then adjusted or written to the parity disks across the entire stripe. With XOR, it doesn't matter whether parity disks exist at all or how many data disks are included in the RAID, since the parity doesn't have to be rewritten when new data disks are added.

In summary:

ZFS writes the logical block, which can be variable in size, to the data disk, including the block checksum. The parity disks are written with the corresponding checksums if there is enough information in the ZIL to completely write the parity stripes, so that no read-write modification occurs and the parity disks do not become hotspots. When the parity disks are written, the atomic transaction for ZFS is complete. This way, in the event of an error, the entire logical block can always be recalculated from the information on the parity disks. The logical ZFS block is then doubly protected: once by the block checksum and once by the RAID checksums on the parity disks. The RAID model is freely expandable with a maximum of three parity disks and a maximum of 32 data disks without having to rewrite all the data. Why 32 data disks? Because with a physical block size of 4K and 32 disks, the stripe size is 128K. In principle, there is no need to set a limit, but according to VDEV, there shouldn't be any more disks either.

**Comment by thulle** [comment]
@NicholasRush  Not sure I'm following correctly, but you seem to be talking about stripes in a way that isn't really how ZFS works? A valid write to a, say, 6 drive Raid-Z3 can be to only 4 drives. If you introduce the concept of full-width stripes you'll always have to do some reading?

**Comment by NicholasRush** [comment]
> [@NicholasRush](https://github.com/NicholasRush) Not sure I'm following correctly, but you seem to be talking about stripes in a way that isn't really how ZFS works? A valid write to a, say, 6 drive Raid-Z3 can be to only 4 drives. If you introduce the concept of full-width stripes you'll always have to do some reading?

@thulle 
No, you misunderstood. I'm sorry, my English isn't that good, so that could be the cause of misunderstandings. A RAID-4 pool without a parity disk is always downgraded and insecure, as it behaves like a RAID-0. For the pool to function, it must have at least one parity disk. However, the parity block of the logical ZFS block is not stored on the parity disks; this is stored normally on the data disks, and then the parity, which is calculated across all the data disks using XOR, is stored on the parity disk. If a logical block cannot be read on the first attempt, the RAID parity must be read on the second attempt to reconstruct and verify the block. The number of parity disks with linear, diagonal, and anti-diagonal parity ensures that up to three hard drives can fail. A ZFS write is complete when the parity disks have also been written.

Simply explained:
It's a different concept than RAID-Z1/2/3. The parity calculation is also conceptually different here. This ensures that additional data disks can be added to the RAID array without the parity on the parity disks or the data on the data disks having to be rewritten. Parity disks can also be removed and added variably. Only the corresponding parity disk then needs to be rewritten.

Why does this work? Because the parity calculation using XOR is always 0 bytes for the corresponding sector if a disk is not present. The 0 bytes does not affect the parity calculation here. If a new data disk is added, it must of course be zeroed accordingly to meet the requirements. To prevent the parity disks from becoming a hotspot, as is usually the case with RAID-4, it must be integrated into ZFS in such a way that the parity is always written synchronously with the payload data. Using copy-on-write, or better yet, redirect-on-write, which ZFS does natively, writing can be synchronized in such a way that this is possible. If you add a new data disk to the RAID 4 array, this disk will receive more writes than the others. What I have in mind isn't a true RAID 4 implementation, but something between JBOD and RAID 0, but with parity disks for data recovery.

**Comment by thulle** [comment]
@NicholasRush I think you might have to make some graphic for how the data is laid out, like this one for a 5 disk RAID-Z:

<img width="312" height="417" alt="Image" src="https://github.com/user-attachments/assets/db3a59de-1801-45d7-9f57-5b53df244ef2" />

Where D is data, P is parity, X is unwriteable due to a minimum of two blocks needed for P+D. The colors indicate different files.

**Comment by NicholasRush** [comment]
@thulle 

Do you mean something like that:

<img width="613" height="234" alt="Image" src="https://github.com/user-attachments/assets/65c2b8d4-0b41-463e-b566-3ba5a2887a6d" />

Light blue is the horizontal stripe for parity calculation. Blue/turquoise is the diagonal parity stripe.

This principle only works if multiple stripes are written in full at once. This eliminates the otherwise necessary reading, modifying, and writing of parity data. The use of the ZFS intent log is therefore mandatory for this RAID principle, as otherwise it will not perform well.

**Comment by thulle** [comment]
Almost, the colors in the image are different files to show how ZFS handles files not covering the full width. You need to handle that.

> This principle only works if multiple stripes are written in full at once.

So, not only does it require full-width stripes, which ZFS doesn't do, it requires multiple full-width stripes at once to be performant?
A single write to A becomes a read of A,B,C,D,E,F,G,H,O,L,P1(ABCD),P1(EFGH)? 12x read amplification?

Also, why is P2 doing parity of P1, but P1 not parity of any P2?

**Comment by NicholasRush** [comment]
@thulle 

Perhaps I should give another example. If each disk has a block size of 4K. And, as in the example, there are four data disks that only hold data and not RAID checksums, then the full stripe that ZFS has to write in this case is 16K. The rules for the logical block size therefore do not change compared to RAID-Z. This model therefore still applies: https://docs.google.com/spreadsheets/d/1_CO8x03VICdiIMulDjQi9NDBd53qFpUreMQVrF1uS28/edit?gid=1224630924#gid=1224630924

However, a logical block being written can also be as small as 4K. Then, in the stripe, 4 x 3 blocks would be 0 bytes. So, 12K bytes would be wasted. To prevent this from happening and ensure the system runs quickly, it is important to ensure that the stripes are sufficiently large and that multiple parity blocks on the parity disk can be updated simultaneously.

However, each data disk can also behave similarly to a JBOD, so a block written can be as small as 4K. Since ZFS is both a volume manager and a RAID system, it is possible to mark these blocks as not wasted. When writing data, ZFS always looks for the largest possible contiguous area. In this case, the data is written in full stripes, similar to RAID 0, across the data disks, and then the parity on the parity disks is updated.

If this is not possible because the VDEV is sufficiently full, the loose blocks that would otherwise be wasted are used, and in this case, the RAID operates more like a JBOD. To update the parity on the parity disks, the parity of the stripes must be recalculated. However, this update of the parity blocks does not occur immediately after the write operation on the parity disks, but only when the ZIL is emptied or full, so that the parity disks do not become hotspots. The parity disks are always written transactionally. This makes it possible to use the best of both worlds: JBOD and RAID-0 for the data disks, with parity disks for data recovery in the event of a failure. Because updating the parity blocks in an emergency is allowed, no blocks are wasted and the logical block size remains variable. Expansion by adding data disks is possible without having to rewrite all the data on the RAID. However, when new data is written, the new data disk receives more data than the other disks, which means that several parity blocks must be rewritten on the parity disks.

This combines the best of both worlds: expandability and flexibility, but also at the expense of performance if no contiguous stripes can be found and many parity blocks need to be updated. The parity blocks that then need to be modified on the parity disks remain in the ZIL until it is cleared so that the parity disks are written as contiguously as possible. The ZIL is extremely important for this, as it is the only way to ensure contiguous file system operations.

Regarding the parity disks themselves:
It is logical that each subsequent parity disk must consider the parity block of the previous disks for calculations, otherwise it would not be possible to restore the first parity disk in the event of a failure while maintaining data integrity. The data integrity of the blocks is realized solely by the parity disks.

**Comment by thulle** [comment]
If you're going to wreck performance and do something that doesn't fit with how ZFS works for the increased flexibility, why do you want this in ZFS and not go for something like unraid parity that can be even more flexible and work with different drive sizes? just slap a writelog onto it if it doesn't have it already?

**Comment by NicholasRush** [comment]
@thulle 

Because the storage principle I'm presenting here has nothing to do with Unraid's, technically speaking.

Unraid isn't designed for the enterprise sector, even though it's also used for storing data there. Furthermore, Unraid is file-based, not block-based.

The storage principle I've presented may have limitations beyond a certain capacity that affect performance. However, this is also the case with RAID-Z and Declustered RAID.

What shouldn't be misunderstood is that the parity disks and the parity stored there only operate in the RAID layer. This also means that no block pointers or checksums need to be rewritten to rewrite parity data. Of course, such a RAID model must be integrated into ZFS in such a way that metadata is written to a different location than the payload, so that both aren't accidentally located on the same data disk.

One RAID construct that will likely be extremely flexible, but also extremely slow, is the HexOS-sponsored OpenZFS AnyRaid. I don't want to speak negatively of it, as they want the same flexibility for data storage as Unraid, but only for ZFS. But what I'm suggesting here will certainly perform better in the end with similar flexibility.

**Comment by NicholasRush** [comment]
@thulle @amotin 

If the RAID-Z model were adapted so that the parity blocks were written to a separate parity disk, then an extension of the VDEV should also be possible. Since RAID-Z already uses XOR for parity calculations, an extension of both the data disks and the parity disks would be possible. My RAID model, which I tried to explain above, would then not have to be implemented.

I'm not trying to have a different conceptual RAID model in ZFS that doesn't work with the fundamentals of ZFS.

I'm only interested in enabling a simple extension of RAID-Z and the ability to remove and add parity drives without having to rewrite all the data. Unfortunately, this is only possible with RAID-Z expansion using rotating parity data. I've read through the "vdev_raidz.c" code. I think it might be possible to redesign the RAID-Z concept so that logical data blocks and parity data blocks are located on separate disks. This would fundamentally change RAID-Z, only in that parity is no longer distributed in rotation across all disks. This, as an additional RAID model, would meet all requirements. So, RAID-Z in the RAID-4 style with dedicated parity disks.

This would allow for extensibility and the addition of data disks to the VDEV, as well as the addition and removal of parity disks from a VDEV at runtime.

It would only be necessary to ensure that when adding parity disks, the second and third parity is possible for the existing logical blocks. And that removing the parity disk by the administrator does not put the pool into a degraded state.

**Comment by amotin** [comment]
@NicholasRush Storing parity on separate disk means exactly one parity block per stripe. Same time RAIDZ allows multiple parity blocks per stripe when storing small blocks.  dRAID changes that concept by indeed storing parity on dedicated disk, just regularly rotating it, but instead it just can't store small blocks without padding to a stripe width.  Attempt to store multiple blocks into one stripe puts already existing data in danger when new ones are written, creating a write hole.  And ZIL has nothing to do with this, since it operates in terms of newly written user data, unrelated to any logic of underlying vdevs.

**Comment by NicholasRush** [comment]
@amotin This means that the only way to implement something like this would be to divide the data disks and parity disks into fixed areas, as with DRaid. But with separate data and parity disks. This would prevent the already written data from being compromised when writing small blocks. This would then result in the logical block size being fixed, resulting in a write hole. Furthermore, this RAID type, like DRaid, would also be suboptimal for writing small data due to the padding. This would ultimately also have to be addressed for DRaid. 

But flexible data storage as I imagine it would then be possible.

I have a question about Raid-Z: Is the ZFS end-to-end block checksum stored in the checksum block or the logical data block? If so, then you wouldn't be able to remove parity disks later. If not, you would have to implement the concept similar to RAID as a normal RAID with a fixed block size and write hole just to achieve this flexibility.

I now understand that the ZIL is only used for incoming user data.

**Comment by jivanpal** [comment]
At the risk of being off-topic: why *does* RAIDz use extent-level parity rather than stripe-level parity? Doing some reading on the topic recently, everyone seems to cite that it's a solution to the write-hole problem, but I don't understand this supposed logic, because:

1. the COW+transaction based nature of ZFS should solve this anyway; and

2. extent-level parity *does not* itself solve the write-hole problem. For example, referring to the diagram [posted above](#issuecomment-3194295656) by @thulle, consider the second (i.e. yellow) file: Suppose that P0, D0, and D2 are written to their respective disks, and then the power is lost. This is a write-hole failure, and to my knowledge/understanding, the only reason that ZFS is able to recover from it is because a transaction containing this file was not committed. That is, the written data blocks D0 and D2 are effectively lost (unless they are recoverable from somewhere else, such as the entire file [D0, D1, D2] having first been written to the ZIL or SLOG before power loss); and if we were to attempt to read the file, the block D1 would not contain the expected data (because it was never written), and thus we would not know which of the blocks [P0, D0, D1, D2] are intact, so cannot recover the file.

Out of everything about ZFS, this detail about how parity is handled for RAIDz datasets truly baffles me — I cannot understand why this design choice was made. As such, an explanation of the benefits of it, and why alternatives like stripe-based parity fail to provide those benefits, would be massively appreciated.

**Comment by NicholasRush** [comment]
@jivanpal 

With ZFS, block parity is implemented only at the logical block level in the file system metadata. Other RAID systems typically have a RAID and a file system layer that complement each other. NetApp's WAFL is one example. And in principle, both ultimately achieve the same goal: end-to-end data integrity. While WAFL was designed more for efficient data storage, this storage efficiency often comes at the expense of performance. However, its performance and storage efficiency are still better than attempting a similar approach with ZFS. Regarding end-to-end data integrity, it's important to note that WAFL only provides disk checksums and RAID checksums. While it lacks the file system tree-based checksum dependency found in ZFS, it still achieves a similar level of behavior.

**Comment by jivanpal** [comment]
@NicholasRush Thanks, but I am asking *why* the decision was made to implement RAIDz's parity in that fashion. For example, you say:

> this storage efficiency often comes at the expense of performance. 

I am interested in e.g. what that performance trade-off is and why it exists, because it is not obvious to me; and whether such a trade-off or other consideration was actually a factor in the decision-making process of the developers. 

Cheers.

**Comment by amotin** [comment]
> Suppose that P0, D0, and D2 are written to their respective disks, and then the power is lost. This is a write-hole failure, and to my knowledge/understanding, the only reason that ZFS is able to recover from it is because a transaction containing this file was not committed.

@jivanpal The hole would be not in the fact that P0, D0 and D2 are lost, but in the P0 shared with D4 and D6, potentially not belonging to this block or even transaction group.  For traditional RAID5 this would be a problem.  But RAIDZ does not share parity between blocks.  This indeed reduces space efficiency of storing very small objects, but instead in case of power loss ZFS can replay the writes from ZIL without caring about recovering unrelated parity that could be damaged. Even more, ZIL replay is not required to even write into the same blocks as was originally.  ZIL just stores "write ZYX into file F", nothing about blocks, parity, etc.

**Comment by jivanpal** [comment]
@amotin Thanks, this makes sense to me now that you've highlighted that the parity sector would be overwritten in place.

However, I initially expected the design to treat each *stripe* (that is, each set of sectors at the same single LBA across the entire array) as a COW unit, meaning parity sectors would never be overwritten in place, just as is the case for any sectors of a regular ZFS (not RAIDz) filesystem. For example, if the yellow file has already been written and committed, and the green file is about to be written, then when the green-coloured D0 is to be written:

1. LBA 2 (consisting of 5 sectors across the entire array) would be copied to memory.
2. The stripe in memory would be altered to include the green D0 at the end (5th sector).
3. The parity sector (yellow P0, 1st sector in memory) would be recalculated from the other 4 sectors in memory, namely [yellow D0, yellow D1, yellow D2, green D0].
4. This entire stripe of 5 sectors in memory would be written out to an unused stripe, so e.g. [new parity, yellow D0, yellow D1, yellow D2, green D0] would be written to LBA 3 or LBA 13 or wherever else is free, but not LBA 2 (which would be an in-place overwrite).
5. Once the remainder of the green file has been written to disk along with relevant parity sectors, the overall change is committed in a new ZFS transaction.

Is this a design that would be of any interest to the OpenZFS team?

**Comment by amotin** [comment]
@jivanpal The problems are on step 4: ZFS does not know to whom the yellow block belong to update the reference.  There can be many references.  Some of the references might be a part of snapshot and can not be updated, so we'd have to keep both copies.  Alternatively we'd have to note somewhere in some table that the blocks previously known as yellow are now stored in different location, that would cost more than all possible benefit from this.

**Comment by jivanpal** [comment]
@amotin Ahhh, thanks, needing to update the references to the yellow data is something I definitely failed to consider. 🙂

**Comment by NicholasRush** [comment]
Now, with the implementation of the AnyRaid feature and the expansion—or rather, the rethinking—of the Block Pointer Model, it might be worth considering the introduction of a general abstraction layer to decouple the physical block address from the ZFS storage address, linking them only indirectly.

This approach creates a range of possibilities: not only does it allow for the mapping of any RAID configuration onto the underlying storage media, but this separation also enables the implementation of features such as deduplication and sub-block compaction, should one choose to do so. Furthermore, storage defragmentation becomes feasible, which is particularly important given that a Zpool can become highly fragmented very quickly following numerous data modifications and the deletion of snapshots.

Ultimately, this would also facilitate a more robust implementation of the encryption feature. This would allow for the encryption and decryption of data without having to modify ZFS's core data model—something that the current Block Pointer structure effectively precludes. However, this would also imply that existing operations—such as compression—could no longer be handled within the current OpenZFS layer; instead, data and checksums would always have to be stored for the *uncompressed* ZFS blocks in order to preserve the current operational behavior of OpenZFS.

In summary, introducing such an abstraction layer would transform current data storage practices and the way implemented RAID models are managed. Furthermore, the existing AnyRAID model would be rendered obsolete, as this approach allows for the emulation of any desired model.
